### DISCUSSION TITLE
Access results from randomized surveys

### DISCUSSION THREAD

**Discussion by psycaroly** [rubenarslan/formr.org#579] [Q&A] [answered]
Dear formr community,

I am using formr for an experimental setup: I use the shuffle function in a run to assign participants randomly to survey A1 or A2.

After this, I have another survey B for everyone, in which  I want to display some (own) answers participants gave in the previous survey. However, given the randomization, I do not know if the answers are stored in survey A1 or A2, so I cannot use survey$variable. I tried ifelse(!is.na(A1$variable), A1$variable, A2$variable), but that didn't work. 

I am guessing this is nothing unusual, so there must be an implemented way to do it, so maybe someone can give me a hint where to look or how to approach this? This would be much appreciated.

I didn't see a similar problem discussed before, so please excuse if I missed it.

Thanks,
Caro

**Comment by rubenarslan** [comment]
Hi Caro, that should work — that's how I would go about it at least. Can you say more about what doesn't work, perhaps provide a reproducible example?

**Comment by psycaroly** [comment]
Hi Ruben,

thanks for the swift reply. 

The survey just doesn't show anything if the randomization didn't shuffle the person in the first group of the ifelse statement, i.e. A1 in my example above, see this screenshot: 
![empty](https://github.com/user-attachments/assets/5856c5c5-ecaa-46ec-b724-a57c4b275913)

If the person _was_ shuffled in the first group, then it shows the result as expected (and if nothing is entered in the target variable, it shows NA, so it's not about the target variable being an empty string),  see this screenshot: 
![not_empty](https://github.com/user-attachments/assets/f89b4359-1e49-464b-ab9b-af85da3cef46)

To be precise, my ifelse()-statement is a nested one, as I have three groups, so it really looks like this, but the logic is the same as in my intitial example:

```
reflection1 <- ifelse(!is.na(survey_A1$reflection1), survey_A1$reflection1, 
                 ifelse(!is.na(survey_A2$reflection1), survey_A2$reflection1, 
                        ifelse(!is.na(survey_A3$reflection1), survey_A3$reflection1, 
                               "Sie haben hier nichts angegeben.")))

cat(reflection1)
```

I tried to make a reproducible example, and attached it: 
[surveytest.json](https://github.com/user-attachments/files/17805617/surveytest.json)
(on pages 2 and 3 the target variables are assessed, and on page 4 they are intended to be shown).

Maybe a workaround could be to access the shuffle$group codings in the survey following the shuffle? Would that work?

Best,
Caro

**Comment by rubenarslan** [comment, marked as answer]
So, if you use debug mode, you can download the attached Rmd file in the last step, where the error occurs.
Then, you can open the code on your local computer and go through it line-by-line.
If you do that, you see that the problem is that surveyA1$reflection1 isn't NA — it's NULL/it doesn't exist. 
So, `heading1 <- ifelse(!is.null(survey_A1$reflection1), "Gemeinsamkeiten", ifelse(!is.null(survey_A2$reflection1), "Unterschieden", "Eigenschaften"))` works.
If you don't want to write that much (quickly becomes confusing), how about `dplyr::coalesce(survey_A1$reflection1, survey_A2$reflection1, survey_A3$reflection1)`.

But yes, you could also refer to the shuffle$group variable in the ifelse condition, that's yet another way.

[formr_rmarkdown (3).Rmd.txt](https://github.com/user-attachments/files/17806225/formr_rmarkdown.3.Rmd.txt)

**Comment by psycaroly** [comment]
Wow, thanks a lot Ruben! Not only for identifying the problem, but also for showing me how I could debug similar problems myself in the future. And for showcasing coalesce, I didn't know that function - so many learnings, it's much appreciated!
